File: trunk/Noxa.Emulation.Psp.Bios.FastHLE/Modules/ThreadManForUser_Time.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>
#include <time.h>

#include "ThreadManForUser.h"
#include "Kernel.h"

using namespace System;
using namespace System::Diagnostics;
using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;
using namespace Noxa::Emulation::Psp::Bios::Modules;

extern int64 _startTick;

// These can all be native
//0x0E927AED _sceKernelReturnFromTimerHandler
//0x110DEC9A sceKernelUSec2SysClock
//0xC8CD158C sceKernelUSec2SysClockWide
//0xBA6B92E2 sceKernelSysClock2USec
//0xE1619D7C sceKernelSysClock2USecWide
//0xDB738F35 sceKernelGetSystemTime
//0x82BC5777 sceKernelGetSystemTimeWide
//0x369ED59D sceKernelGetSystemTimeLow

// SysTime is low, high

// void _sceKernelReturnFromTimerHandler(); (/user/pspthreadman.h:1447)
void ThreadManForUser::_sceKernelReturnFromTimerHandler(){}

// int sceKernelUSec2SysClock(unsigned int usec, SceKernelSysClock *clock); (/user/pspthreadman.h:1463)
int ThreadManForUser::sceKernelUSec2SysClock( IMemory^ memory, int usec, int clock )
{
	// sysclock is us count, so this is easy
	memory->WriteWord( clock, 4, usec );
	memory->WriteWord( clock + 4, 4, 0 );

	return 0;
}

// SceInt64 sceKernelUSec2SysClockWide(unsigned int usec); (/user/pspthreadman.h:1472)
int64 ThreadManForUser::sceKernelUSec2SysClockWide( int usec )
{
	// sysclock is us count, so this is easy
	return ( int64 )( uint )usec;
}

// int sceKernelSysClock2USec(SceKernelSysClock *clock, unsigned int *sec, unsigned int *usec); (/user/pspthreadman.h:1483)
int ThreadManForUser::sceKernelSysClock2USec( IMemory^ memory, int clock, int sec, int usec )
{
	// sysclock is us count - we need to get seconds + usec
	int clo = memory->ReadWord( clock );
	int chi = memory->ReadWord( clock + 4 );
	int64 c = memory->ReadDoubleWord( clock );

	memory->WriteWord( sec, 4, ( int )( c / 1000000 ) );
	memory->WriteWord( usec, 4, ( int )( c % 1000000 ) );
	
	return 0;
}

// int sceKernelSysClock2USecWide(SceInt64 clock, unsigned *sec, unsigned int *usec); (/user/pspthreadman.h:1494)
int ThreadManForUser::sceKernelSysClock2USecWide( IMemory^ memory, int64 clock, int sec, int usec )
{
	memory->WriteWord( sec, 4, ( int )( clock / 1000000 ) );
	memory->WriteWord( usec, 4, ( int )( clock % 1000000 ) );
	
	return 0;
}

// int sceKernelGetSystemTime(SceKernelSysClock *time); (/user/pspthreadman.h:1503)
int ThreadManForUser::sceKernelGetSystemTime( IMemory^ memory, int time )
{
	// microseconds since game started
	int64 ticks = DateTime::Now.Ticks - _kernel->StartTick;
	// 10000000 ticks/sec - 1000000 us/sec
	ticks /= 10;

	int lo = ( int )ticks;
	int hi = ( int )( ticks >> 32 );

	memory->WriteWord( time, 4, lo );
	memory->WriteWord( time + 4, 4, hi );

	return 0;
}

// SceInt64 sceKernelGetSystemTimeWide(); (/user/pspthreadman.h:1510)
int64 ThreadManForUser::sceKernelGetSystemTimeWide()
{
	// microseconds since game started
	int64 ticks = DateTime::Now.Ticks - _kernel->StartTick;
	return ticks / 10;
}

#pragma unmanaged
int sceKernelGetSystemTimeLowN()
{
	ULARGE_INTEGER time;
	GetSystemTimeAsFileTime( ( FILETIME* )&time );
	return ( int )( ( time.QuadPart - _startTick ) / 10 );

}
#pragma managed

// unsigned int sceKernelGetSystemTimeLow(); (/user/pspthreadman.h:1517)
int ThreadManForUser::sceKernelGetSystemTimeLow()
{
	// microseconds since game started
	int64 ticks = DateTime::Now.Ticks - _kernel->StartTick;
	ticks /= 10;
	return ( int )ticks;
}

File: Noxa.Emulation.Psp.Bios.ManagedHLE/Modules/ThreadManForUser.Time.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

using Noxa.Utilities;
using Noxa.Emulation.Psp;
using Noxa.Emulation.Psp.Bios;
using Noxa.Emulation.Psp.Cpu;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE.Modules
{
	partial class ThreadManForUser
	{
		//[NotImplemented]
		//[Stateless]
		//[BiosFunction( 0x0E927AED, "_sceKernelReturnFromTimerHandler" )]
		//// SDK location: /user/pspthreadman.h:1447
		//// SDK declaration: void _sceKernelReturnFromTimerHandler();
		//public void _sceKernelReturnFromTimerHandler()
		//{
		//}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x110DEC9A, "sceKernelUSec2SysClock" )]
		// SDK location: /user/pspthreadman.h:1463
		// SDK declaration: int sceKernelUSec2SysClock(unsigned int usec, SceKernelSysClock *clock);
		public int sceKernelUSec2SysClock( int usec, int clock )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xC8CD158C, "sceKernelUSec2SysClockWide" )]
		// SDK location: /user/pspthreadman.h:1472
		// SDK declaration: SceInt64 sceKernelUSec2SysClockWide(unsigned int usec);
		public long sceKernelUSec2SysClockWide( int usec )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xBA6B92E2, "sceKernelSysClock2USec" )]
		// SDK location: /user/pspthreadman.h:1483
		// SDK declaration: int sceKernelSysClock2USec(SceKernelSysClock *clock, unsigned int *low, unsigned int *high);
		public int sceKernelSysClock2USec( int clock, int low, int high )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xE1619D7C, "sceKernelSysClock2USecWide" )]
		// SDK location: /user/pspthreadman.h:1494
		// SDK declaration: int sceKernelSysClock2USecWide(SceInt64 clock, unsigned *low, unsigned int *high);
		public int sceKernelSysClock2USecWide( long clock, int low, int high )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xDB738F35, "sceKernelGetSystemTime" )]
		// SDK location: /user/pspthreadman.h:1503
		// SDK declaration: int sceKernelGetSystemTime(SceKernelSysClock *time);
		public int sceKernelGetSystemTime( int time )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x82BC5777, "sceKernelGetSystemTimeWide" )]
		// SDK location: /user/pspthreadman.h:1510
		// SDK declaration: SceInt64 sceKernelGetSystemTimeWide();
		public long sceKernelGetSystemTimeWide()
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x369ED59D, "sceKernelGetSystemTimeLow" )]
		// SDK location: /user/pspthreadman.h:1517
		// SDK declaration: unsigned int sceKernelGetSystemTimeLow();
		public int sceKernelGetSystemTimeLow()
		{
			return Module.NotImplementedReturn;
		}
	}
}
